### PR TITLE
Fix for rule based labeling toSld()

### PR DESCRIPTION
**Pull request by mhugo**
## Description
This fixes QgsRuleBasedLabeling::toSld() when a rule has a null setting. Corresponding unit test updated.

I've also replaced some deletes in QgsRuleBasedLabeling with smart pointers.

@pblottiere I would be glad if you could have a look :)

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/core/qgsrulebasedlabeling.h [hunk 1]
```diff
@@ -325,7 +325,7 @@ class CORE_EXPORT QgsRuleBasedLabeling : public QgsAbstractVectorLayerLabeling
 
       private:
         Rule *mParent; // parent rule (NULL only for root rule)
-        QgsPalLayerSettings *mSettings = nullptr;
+        QScopedPointer<QgsPalLayerSettings> mSettings;
```
Comment (nyalldawson) [inline src/core/qgsrulebasedlabeling.h]: Use std::unique_ptr instead of the unmaintained qt smart pointers

Comment (pblottiere) [reply src/core/qgsrulebasedlabeling.h]: +1

Comment (mhugo) [reply src/core/qgsrulebasedlabeling.h]: Oh yes ! I forgot we can use unique_ptr now. Good ! and thanks :)

File: src/core/qgsrulebasedlabeling.h [hunk 2]
```diff
@@ -336,8 +336,7 @@ class CORE_EXPORT QgsRuleBasedLabeling : public QgsAbstractVectorLayerLabeling
 
         QString mRuleKey; // string used for unique identification of rule within labeling
 
-        // temporary
-        QgsExpression *mFilter = nullptr;
+        QScopedPointer<QgsExpression> mFilter;
```
Comment (m-kuhn) [inline src/core/qgsrulebasedlabeling.h]: Same here? ;)

**Comment by domlysz** [comment]
It seems this pull request cause regression with data defined label position because of missing uuid for labeling rules.

https://issues.qgis.org/issues/19979

**Pull request merged**